numpy/_core/src/umath/loops.c.src

```diff
@@ -474,6 +474,38 @@ NPY_NO_EXPORT NPY_GCC_OPT_3 int
 NPY_NO_EXPORT void
 @TYPE@_power(char **args, npy_intp const *dimensions, npy_intp const *steps, void *NPY_UNUSED(func))
 {
+    if (steps[1]==0) {
+        // stride for second argument is 0
+        BINARY_DEFS_ZERO_STRIDE
+        const @type@ in2 = *(@type@ *)ip2;
+        #if @SIGNED@
+                if (in2 < 0) {
+                    npy_gil_error(PyExc_ValueError,
+                                  "Integers to negative integer powers are not allowed.");
+                    return;
+                }
+        #endif
+
+        int first_bit = in2 & 1;
+        @type@ in2start = in2 >> 1;
+
+        BINARY_LOOP_SLIDING_ZERO_STRIDE {
+            @type@ in1 = *(@type@ *)ip1;
+            @type@ var = in2start;
+            @type@ out;
+
+            out = first_bit ? in1 : 1;
+            while (var > 0) {
+                in1 *= in1;
+                if (var & 1) {
+                    out *= in1;
+                }
+                var >>= 1;
+            }
+            *((@type@ *) op1) = out;
+        }
+        return;
+    }
```

Comment:
I do wonder if we go here, if we can't create a small helper or so and not duplicate the code at all (i.e. this is just the identical code, I think.  The compiler should be able to do the same optimizations?)

Reply:
It is indeed almost identical to the code on lines 532 to 539 (the only difference is the name of `var` and `x2`). We could create an method for this piece of code, probably the compiler will inline. If we want to be sure we have to create a macro. Is that what you mean?

Reply:
Yeah we had a brief look yesterday and there was some confusion about added complexity so not repeating the code seemed like a good idea.  I would expect a static inline function should be sufficient in practice.